plugins/wallabag.koplugin/main.lua

```diff
@@ -0,0 +1,695 @@
+                        keep_menu_open = true,
+                        callback = function()
+                            UIManager:show(InfoMessage:new{
+                                text = _([[Download directory: use a directory that is exclusively used by the Wallabag plugin. Existing files in this directory risk being deleted.
+
+Articles marked as finished or 100% read can be deleted from the server. Those articles can also be deleted automatically when downloading new articles if the 'Process deletions during download' option is enabled.
+
+The 'Synchronise remotely delete files' option will remove local files that do not exist anymore on the server.]])
+                            })
+                        end,
+                    }
+                }
+            },
+            {
+                text = _("Info"),
+                keep_menu_open = true,
+                callback = function()
+                    UIManager:show(InfoMessage:new{
+                        text = T(_([[Wallabag is an open source Read-it-later service. This plugin synchronises with a Wallabag server.
+
+More details: https://wallabag.org
+
+Downloads to directory: %1]]), filemanagerutil.abbreviate(self.directory))
+                    })
+                end,
+            },
+        },
+    }
+end
+
+function Wallabag:getBearerToken()
+
+    -- Check if the configuration is complete
+    local function isempty(s)
+        return s == nil or s == ""
+    end
+
+    if isempty(self.server_url) or isempty(self.username) or isempty(self.password)  or isempty(self.client_id) or isempty(self.client_secret) or isempty(self.directory) then
+        UIManager:show(InfoMessage:new{
+            text = _('Please configure the server and local settings.')
+        })
+        return false
+    end
+
+    -- Check if the download directory is valid
+    local dir_mode = lfs.attributes(self.directory, "mode")
+    logger.dbg("mode:", dir_mode)
+    if dir_mode ~= "directory" then
+         UIManager:show(InfoMessage:new{
+            text = _('The download directory is not valid.\nPlease configure it in the settings.')
+        })
+        return false
+    end
+    if string.sub( self.directory, -1 ) ~= '/' then
+        self.directory = self.directory .. '/'
+    end
+
+    local now = os.time()
+    if self.token_expiry - now > 300 then
+        -- token still valid for a while, no need to renew
+        return true
+    end
+
+    local login_url = "/oauth/v2/token"
+    local body = "{ \"grant_type\": \"password\", \"client_id\": \"" .. self.client_id .. "\", \"client_secret\": \"" .. self.client_secret .. "\", \"username\": \"" .. self.username .. "\", \"password\": \"" .. self.password .. "\"}"
+    local headers = {
+        ["Content-type"] = "application/json",
+        ["Accept"] = "application/json, */*",
+        ["Content-Length"] = tostring(#body),
+        }
+    local result = self:callAPI( 'POST', login_url, headers, body, "" )
```

Comment:
I have to wonder actually, is there a specific rationale behind where you used single and where double quotes?

Reply:
I'm afraid there was no specific logic in my use of quotes.
I starting by grabbing a few bits of interesting code in other plugins and after a while the inconsistencies spread a bit.

All changed to double quotes and a few multilines.